Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,785 @@
+
+    -- | Call @stack exec ghc-pkg@ to find the package the provides a module.
+    -- stackGhcPkgFindModule :: String -> IO (Maybe String)
+    stackGhcPkgFindModule m = do
+        let opts = ["exec", "ghc-pkg", "find-module", m, "--", "--simple-output"]
+
+        x <- liftIO $ executeFallibly' "stack" opts
+
+        case x of
+            Nothing             -> return Nothing
+            Just (output, err)  -> do gmLog GmDebug "" $ strDoc $ "stackGhcPkgFindModule stdout: " ++ show output
+                                      gmLog GmDebug "" $ strDoc $ "stackGhcPkgFindModule stderr: " ++ show err
+                                      return $ join $ (Safe.lastMay . words) <$> (Safe.lastMay . lines) output
+
+ghcPkgHaddockUrl
+    :: forall m. (GmLog m, GmOut m, MonadIO m)
+    => FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> String
+    -> m (Maybe String)
+ghcPkgHaddockUrl ghcPkg readProc pkgDbStack p = do
+    gmLog GmDebug "" $ strDoc $ "ghcPkgHaddockUrl: " ++ p
+
+    let p' = case splitOn "@" p of
+                [p0, _] -> p0
+                _       -> p
+
+    hout <- liftIO $ readProc ghcPkg (toDocDirOpts p' pkgDbStack) ""
+    return $ Safe.lastMay $ words $ reverse . dropWhile (== '\n') . reverse $ hout
+  where
+    -- This fails unless we have --global and --user, unlike
+    -- pkgDoc elsewhere in ghc-mod.
+    toDocDirOpts pkg dbs = ["field", pkg, "haddock-html", "--global", "--user"] ++ ghcPkgDbStackOpts dbs
+
+ghcPkgHaddockInterface
+    :: forall (m :: * -> *).  MonadIO m
+    => FilePath
+    -> (FilePath -> [String] -> String -> IO String)
+    -> [GhcPkgDb]
+    -> String
+    -> m (Maybe String)
+ghcPkgHaddockInterface ghcPkg readProc pkgDbStack p = do
+    hout <- liftIO $ readProc ghcPkg (toHaskellInterfaces p pkgDbStack) ""
+    return $ Safe.lastMay $ words $ reverse . dropWhile (== '\n') . reverse $ hout
+  where
+    toHaskellInterfaces pkg dbs = ["field", pkg, "haddock-interfaces", "--global", "--user"] ++ ghcPkgDbStackOpts dbs
+
+getVisibleExports
+    :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m)
+    => (String -> m (Maybe String))
+    -> String
+    -> m (Maybe (M.Map String [String]))
+getVisibleExports getHaddockInterfaces p = do
+    gmLog GmDebug "" $ strDoc $ "getVisibleExports: " ++ p
+
+    let p' = case splitOn "@" p of
+                [p0, _] -> p0
+                _       -> p
+    
+    haddockInterfaceFile <- getHaddockInterfaces p'
+    join <$> traverse getVisibleExports' haddockInterfaceFile
+
+  where
+
+    getVisibleExports' :: forall m. (GhcMonad m, MonadIO m, GmOut m, GmLog m) => FilePath -> m (Maybe (M.Map String [String]))
+    getVisibleExports' ifile = do
+        iface <- Haddock.readInterfaceFile nameCacheFromGhc ifile
+
+        case iface of
+            Left _          -> do gmErrStrLn $ "Failed to read the Haddock interface file: " ++ ifile
```

Comment:
You should use the `MonadError GhcModError` for errors, just add a new constructor to `GhcModError` and add a pretty printer to `gmeDoc`.